docs/SharedBrokerPool.md

```diff
@@ -0,0 +1,43 @@
+---
+id: shared-broker-pool
+title: Sharing Broker Pools
+---
+
+The default behavior of KafkaJS is to treat each consumer, producer, and admin client independently from each other. In particular this means network connections to the configured brokers will be created again for every one of these. For larger applications this can lead to high resource usage both on the client side as well as on the broker side, and it may lead to throttling.
+
+It is possible to influence this behavior for consumers and producers by manually configuring "broker pools", and sharing these.
+
+## <a name="consumer"></a> Example
+
+```javascript
+const brokerPool = kafka.brokerPool({ })
+await brokerPool.connect()
+```
+
+This broker pool can be provided to the `kafka.consumer` and `kafka.producer` methods, instead of the options that would typically configure the pool.
+
+```javascript
+const consumer = kafka.consumer({ groupId: 'my-group', brokerPool })
+```
+
+```javascript
+const producer = kafka.producer({ brokerPool })
+```
+
+The consumer and producer can then be used as described in [Consuming Messages](Consuming.md) and [Producing Messages](Producing.md), with the caveat that the `connect` and `disconnect` methods will affect the complete pool, and therefore should typicaly not be used.
```

Comment:
How will graceful shutdowns work with this? Let's say I have a typical stream processor, consuming from one topic and publishing into another.

In a normal KafkaJS application, I'd want to first shut down the consumer and then the producer, so that while my consumer is in the process of shutting down the producer is still up to be able to be used while wrapping up. How would that work with this solution?

Also, is there a way we can prevent people from shooting themselves in the foot here, since now calling `consumer.disconnect()` could have unintended consequences if you're not paying attention. I'm not sure what the best option is, to be honest.

Reply:
> Also, is there a way we can prevent people from shooting themselves in the foot here, since now calling consumer.disconnect() could have unintended consequences if you're not paying attention. I'm not sure what the best option is, to be honest.

I toyed with this a bit a few minutes ago, exactly because I shot myself in the foot when the crash handling disconnected my cluster :)

I eventually decided that the crash handler definitely was wrong in attempting it, and pushed the problem away for now. I find it difficult to have names like "connect" and "disconnect" on the consumer/producers, because there's not much there except doing exactly that. Having connect/disconnect not do connect/disconnect is just simply confusing. But, at the same time disconnecting a shared pool a real issue. Still thinking as well, but I could see an approach that would return a different class depending on whether you have or have not provided a shared pool, which should be possible to model with types so that the compiler can scream at you.

Reply:
> Still thinking as well, but I could see an approach that would return a different class depending on whether you have or have not provided a shared pool, which should be possible to model with types so that the compiler can scream at you.

This is what I had in mind as well, for exactly the same reasons you described.

Reply:
I asked for some feedback from internal customers of KafkaJS on this API, to see if they had some thoughts, and one person had a very interesting idea that I think makes a lot of sense, even if it makes life a bit more annoying for us.

What if we don't expose `connect` and `disconnect` on the connection pool at all, but rather have the connection pool connect if it's not connected, and disconnect when there are no more clients that require it to be connected?

For example:

```js
const connectionPool = kafka.connectionPool()
const consumer = kafka.consumer({ connectionPool })
const producer = kafka.producer({ connectionPool })

// ConnectionPool is not connected, so it connects
await producer.connect()

// Consumer registers that it also wants to be connected
await consumer.connect()

// Consumer doesn't want to be connected anymore.
// It stops etc, but the connection pool is still connected.
await consumer.disconnect()

// No more connections are required, connection pool disconnects.
await producer.disconnect()
```

That way the difference in the application code is almost 0 whether you have connection pooling or not. As a user, the way you describe the dependency between instances of producers and consumers is explicit and clear. The only downside I see is that the connection pool has to do a bit of work to keep track of when it should disconnect and not.

Reply:
I think this is a good idea.

Essentially the pool knows about it's consumers at this point, which would produce a cycle -- but maybe a simple counter of how many consumers exist might be enough.

Reply:
I think https://github.com/tulios/kafkajs/pull/977/commits/ccfaa69cef0f8741cf5af122ab9c268402c495e8 actually does that quite nicely (to be squashed into the main commit eventually, but now pulled out to show it)